README.md
rnoaa
=====



[![Build Status](https://api.travis-ci.org/ropensci/rnoaa.png)](https://travis-ci.org/ropensci/rnoaa)
[![Build status](https://ci.appveyor.com/api/projects/status/8daqtllo2sg6me07/branch/master)](https://ci.appveyor.com/project/sckott/rnoaa/branch/master)
[![Coverage Status](https://coveralls.io/repos/ropensci/rnoaa/badge.svg)](https://coveralls.io/r/ropensci/rnoaa)
[![cran version](http://www.r-pkg.org/badges/version/rnoaa)](http://cran.rstudio.com/web/packages/rnoaa)

## Help

There is a tutorial on the [rOpenSci website](http://ropensci.org/tutorials/rnoaa_tutorial.html), and there are many tutorials in the package itself, available in your R session, or [on CRAN](http://cran.r-project.org/web/packages/rnoaa/index.html). The tutorials:

* NOAA Buoy vignette
* NOAA National Climatic Data Center (NCDC) vignette (examples)
* NOAA NCDC attributes vignette
* NOAA NCDC workflow vignette
* Sea ice vignette
* Severe Weather Data Inventory (SWDI) vignette
* Historical Observing Metadata Repository (HOMR) vignette
* Storms (IBTrACS) vignette

## Data sources used in rnoaa

* NOAA NCDC climate data:
    * We are using the NOAA API version 2
    * Docs for the NCDC API are at http://www.ncdc.noaa.gov/cdo-web/webservices/v2
    * GHCN Daily data is available at http://www.ncdc.noaa.gov/oa/climate/ghcn-daily/ via FTP and HTTP
* Severe weather data docs are at http://www.ncdc.noaa.gov/swdiws/
* [Sea ice data](ftp://sidads.colorado.edu/DATASETS/NOAA/G02135/shapefiles)
* [NOAA buoy data](http://www.ndbc.noaa.gov/)
* [ERDDAP data](http://upwell.pfeg.noaa.gov/erddap/index.html)
  * Now in package [rerddap](https://github.com/ropensci/rerddap)
* Tornadoes! Data from the [NOAA Storm Prediction Center](http://www.spc.noaa.gov/gis/svrgis/)
* HOMR - Historical Observing Metadata Repository - from [NOAA NCDC](http://www.ncdc.noaa.gov/homr/api)
* Storm data - from the [International Best Track Archive for Climate Stewardship (IBTrACS)](http://www.ncdc.noaa.gov/ibtracs/index.php?name=wmo-data)
* [GHCND FTP data](ftp://ftp.ncdc.noaa.gov/pub/data/noaa) - NOAA NCDC API has some/all (not sure really) of this data, but FTP allows to get more data more quickly

## netcdf data

Functions to work with buoy data use netcdf files. You'll need the `ncdf` package for those functions, and those only. `ncdf` is in Suggests in this package, meaning you only need `ncdf` if you are using the buoy functions. You'll get an informative error telling you to install `ncdf` if you don't have it and you try to use the buoy functions. Installation of `ncdf` should be straightforward on Mac and Windows, but on Linux you may have issues. See http://cran.r-project.org/web/packages/ncdf/INSTALL

## NOAA NCDC Datasets

There are many NOAA NCDC datasets. All data sources work, except `NEXRAD2` and `NEXRAD3`, for an unkown reason.

| Dataset | Description | Start date | End date |
|---------|-------------|------------|----------|
| ANNUAL | Annual Summaries | 1831-02-01 | 2013-11-01 |
| GHCND | Daily Summaries | 1763-01-01 | 2014-03-15 |
| GHCNDMS | Monthly Summaries | 1763-01-01 | 2014-01-01 |
| NORMAL_ANN | Normals Annual/Seasonal | 2010-01-01 | 2010-01-01 |
| NORMAL_DLY | Normals Daily | 2010-01-01 | 2010-12-31 |
| NORMAL_HLY | Normals Hourly | 2010-01-01 | 2010-12-31 |
| NORMAL_MLY | Normals Monthly | 2010-01-01 | 2010-12-01 |
| PRECIP_15 | Precipitation 15 Minute | 1970-05-12 | 2013-03-01 |
| PRECIP_HLY | Precipitation Hourly | 1900-01-01 | 2013-03-01 |
| NEXRAD2 | Nexrad Level II | 1991-06-05 | 2014-03-14 |
| NEXRAD3 | Nexrad Level III | 1994-05-20 | 2014-03-11 |

## NOAA NCDC Attributes

Each NOAA dataset has a different set of attributes that you can potentially get back in your search. See http://www.ncdc.noaa.gov/cdo-web/datasets for detailed info on each dataset. We provide some information on the attributes in this package; see the [vignette for attributes](inst/vign/rncdc_attributes.md) to find out more

## Authentication

You'll need an API key to use the NOAA NCDC functions (those starting with `ncdc*()`) in this package (essentially a password). Go to http://www.ncdc.noaa.gov/cdo-web/token to get one. *You can't use this package without an API key.*

Once you obtain a key, there are two ways to use it.

a) Pass it inline with each function call (somewhat cumbersome)


```r
ncdc(datasetid = 'PRECIP_HLY', locationid = 'ZIP:28801', datatypeid = 'HPCP', limit = 5, token =  "YOUR_TOKEN")
```

b) Alternatively, you might find it easier to set this as an option, either by adding this line to the top of a script or somewhere in your `.rprofile`


```r
options(noaakey = "KEY_EMAILED_TO_YOU")
```

c) You can always store in permamently in your `.Rprofile` file.


## Installation

__GDAL__

You'll need [GDAL](http://www.gdal.org/) installed first. You may want to use GDAL >= `0.9-1` since that version or later can read TopoJSON format files as well, which aren't required here, but may be useful. Install GDAL:

* OSX - From http://www.kyngchaos.com/software/frameworks
* Linux - run `sudo apt-get install gdal-bin` [reference](https://www.mapbox.com/tilemill/docs/guides/gdal/#linux)
* Windows - From http://trac.osgeo.org/osgeo4w/

Then when you install the R package `rgdal` (`rgeos` also requires GDAL), you'll most likely need to specify where you're `gdal-config` file is on your machine, as well as a few other things. I have an OSX Mavericks machine, and this works for me (there's no binary for Mavericks, so install the source version):


```r
install.packages("http://cran.r-project.org/src/contrib/rgdal_0.9-1.tar.gz", repos = NULL, type="source", configure.args = "--with-gdal-config=/Library/Frameworks/GDAL.framework/Versions/1.10/unix/bin/gdal-config --with-proj-include=/Library/Frameworks/PROJ.framework/unix/include --with-proj-lib=/Library/Frameworks/PROJ.framework/unix/lib")
```

The rest of the installation should be easy. If not, let us know.

__Stable version from CRAN__


```r
install.packages("rnoaa")
```

__or development version from GitHub__


```r
install.packages("devtools")
devtools::install_github("ropensci/rnoaa")
library('rnoaa')
```

__or version with buoy functions on Github__


```r
install.packages("devtools")
devtools::install_github("ropensci/rnoaa", ref="buoy")
```


```r
library('rnoaa')
```

## NCDC v2 API data

###  Fetch list of city locations in descending order


```r
ncdc_locs(locationcategoryid='CITY', sortfield='name', sortorder='desc')
#> $meta
#> $meta$totalCount
#> [1] 1657
#> 
#> $meta$pageCount
#> [1] 25
#> 
#> $meta$offset
#> [1] 1
#> 
#> 
#> $data
#>       mindate    maxdate                  name datacoverage            id
#> 1  1892-08-01 2015-05-31            Zwolle, NL       1.0000 CITY:NL000012
#> 2  1901-01-01 2015-06-15            Zurich, SZ       1.0000 CITY:SZ000007
#> 3  1906-01-01 1980-12-31            Zinder, NG       0.8678 CITY:NG000004
#> 4  1938-01-01 2015-06-15         Zhytomyra, UP       0.9731 CITY:UP000025
#> 5  1948-03-01 2015-06-15        Zhezkazgan, KZ       0.9294 CITY:KZ000017
#> 6  1951-01-01 2015-06-15         Zhengzhou, CH       1.0000 CITY:CH000045
#> 7  1941-01-01 2014-12-22          Zaragoza, SP       1.0000 CITY:SP000021
#> 8  1936-01-01 2009-06-16      Zaporiyhzhya, UP       0.9739 CITY:UP000024
#> 9  1893-01-01 2015-06-18     Zanesville, OH US       1.0000 CITY:US390029
#> 10 1912-01-01 1971-12-31             Zahle, LE       0.7972 CITY:LE000004
#> 11 1951-01-01 2010-05-19           Zahedan, IR       0.9930 CITY:IR000019
#> 12 1860-12-01 2013-12-31            Zagreb, HR       1.0000 CITY:HR000002
#> 13 1947-01-01 2015-05-30 Yuzhno-Sakhalinsk, RS       1.0000 CITY:RS000081
#> 14 1893-01-01 2015-06-18           Yuma, AZ US       1.0000 CITY:US040015
#> 15 1942-02-01 2015-06-18   Yucca Valley, CA US       1.0000 CITY:US060048
#> 16 1893-01-01 2015-06-18      Yuba City, CA US       1.0000 CITY:US060047
#> 17 1893-01-01 2015-06-18     Youngstown, OH US       1.0000 CITY:US390028
#> 18 1941-01-01 2015-06-18           York, PA US       1.0000 CITY:US420024
#> 19 1876-01-01 2015-06-18        Yonkers, NY US       1.0000 CITY:US360031
#> 20 1949-01-01 2015-06-15          Yokohama, JA       1.0000 CITY:JA000017
#> 21 1951-01-01 2015-06-15          Yinchuan, CH       1.0000 CITY:CH000044
#> 22 1885-06-01 2015-06-15           Yerevan, AM       0.9751 CITY:AM000001
#> 23 1948-01-01 2015-06-18     Yazoo City, MS US       1.0000 CITY:US280020
#> 24 1959-07-01 1987-05-20         Yaroslavl, RS       0.9850 CITY:RS000080
#> 25 1932-01-01 2015-06-18        Yankton, SD US       1.0000 CITY:US460009
#> 
#> attr(,"class")
#> [1] "ncdc_locs"
```

### Get info on a station by specifcying a dataset, locationtype, location, and station


```r
ncdc_stations(datasetid='GHCND', locationid='FIPS:12017', stationid='GHCND:USC00084289')
#> $meta
#> NULL
#> 
#> $data
#>   elevation    mindate    maxdate latitude                  name
#> 1      12.2 1899-02-01 2015-06-16  28.8029 INVERNESS 3 SE, FL US
#>   datacoverage                id elevationUnit longitude
#> 1            1 GHCND:USC00084289        METERS  -82.3126
#> 
#> attr(,"class")
#> [1] "ncdc_stations"
```


### Search for data


```r
out <- ncdc(datasetid='NORMAL_DLY', stationid='GHCND:USW00014895', datatypeid='dly-tmax-normal', startdate = '2010-05-01', enddate = '2010-05-10')
```

### See a data.frame


```r
head( out$data )
#>                  date        datatype           station value fl_c
#> 1 2010-05-01T00:00:00 DLY-TMAX-NORMAL GHCND:USW00014895   652    S
#> 2 2010-05-02T00:00:00 DLY-TMAX-NORMAL GHCND:USW00014895   655    S
#> 3 2010-05-03T00:00:00 DLY-TMAX-NORMAL GHCND:USW00014895   658    S
#> 4 2010-05-04T00:00:00 DLY-TMAX-NORMAL GHCND:USW00014895   661    S
#> 5 2010-05-05T00:00:00 DLY-TMAX-NORMAL GHCND:USW00014895   663    S
#> 6 2010-05-06T00:00:00 DLY-TMAX-NORMAL GHCND:USW00014895   666    S
```

### Plot data, super simple, but it's a start


```r
out <- ncdc(datasetid='GHCND', stationid='GHCND:USW00014895', datatypeid='PRCP', startdate = '2010-05-01', enddate = '2010-10-31', limit=500)
ncdc_plot(out, breaks="1 month", dateformat="%d/%m")
```

![plot of chunk unnamed-chunk-13](inst/img/unnamed-chunk-13-1.png) 

### More plotting

You can pass many outputs from calls to the `noaa` function in to the `ncdc_plot` function.


```r
out1 <- ncdc(datasetid='GHCND', stationid='GHCND:USW00014895', datatypeid='PRCP', startdate = '2010-03-01', enddate = '2010-05-31', limit=500)
out2 <- ncdc(datasetid='GHCND', stationid='GHCND:USW00014895', datatypeid='PRCP', startdate = '2010-09-01', enddate = '2010-10-31', limit=500)
ncdc_plot(out1, out2, breaks="45 days")
```

![plot of chunk unnamed-chunk-14](inst/img/unnamed-chunk-14-1.png) 

### Get table of all datasets


```r
ncdc_datasets()
#> $meta
#> $meta$offset
#> [1] 1
#> 
#> $meta$count
#> [1] 11
#> 
#> $meta$limit
#> [1] 25
#> 
#> 
#> $data
#>                     uid    mindate    maxdate                      name
#> 1  gov.noaa.ncdc:C00040 1831-02-01 2015-01-01          Annual Summaries
#> 2  gov.noaa.ncdc:C00861 1763-01-01 2015-06-17           Daily Summaries
#> 3  gov.noaa.ncdc:C00841 1763-01-01 2015-05-01         Monthly Summaries
#> 4  gov.noaa.ncdc:C00345 1991-06-05 2015-06-18  Weather Radar (Level II)
#> 5  gov.noaa.ncdc:C00708 1994-05-20 2015-06-14 Weather Radar (Level III)
#> 6  gov.noaa.ncdc:C00821 2010-01-01 2010-01-01   Normals Annual/Seasonal
#> 7  gov.noaa.ncdc:C00823 2010-01-01 2010-12-31             Normals Daily
#> 8  gov.noaa.ncdc:C00824 2010-01-01 2010-12-31            Normals Hourly
#> 9  gov.noaa.ncdc:C00822 2010-01-01 2010-12-01           Normals Monthly
#> 10 gov.noaa.ncdc:C00505 1970-05-12 2014-01-01   Precipitation 15 Minute
#> 11 gov.noaa.ncdc:C00313 1900-01-01 2014-01-01      Precipitation Hourly
#>    datacoverage         id
#> 1          1.00     ANNUAL
#> 2          1.00      GHCND
#> 3          1.00    GHCNDMS
#> 4          0.95    NEXRAD2
#> 5          0.95    NEXRAD3
#> 6          1.00 NORMAL_ANN
#> 7          1.00 NORMAL_DLY
#> 8          1.00 NORMAL_HLY
#> 9          1.00 NORMAL_MLY
#> 10         0.25  PRECIP_15
#> 11         1.00 PRECIP_HLY
#> 
#> attr(,"class")
#> [1] "ncdc_datasets"
```

### Get data category data and metadata


```r
ncdc_datacats(locationid='CITY:US390029')
#> $meta
#> $meta$totalCount
#> [1] 37
#> 
#> $meta$pageCount
#> [1] 25
#> 
#> $meta$offset
#> [1] 1
#> 
#> 
#> $data
#>                     name            id
#> 1    Annual Agricultural        ANNAGR
#> 2     Annual Degree Days         ANNDD
#> 3   Annual Precipitation       ANNPRCP
#> 4     Annual Temperature       ANNTEMP
#> 5    Autumn Agricultural         AUAGR
#> 6     Autumn Degree Days          AUDD
#> 7   Autumn Precipitation        AUPRCP
#> 8     Autumn Temperature        AUTEMP
#> 9               Computed          COMP
#> 10 Computed Agricultural       COMPAGR
#> 11           Degree Days            DD
#> 12      Dual-Pol Moments DUALPOLMOMENT
#> 13             Echo Tops       ECHOTOP
#> 14      Hydrometeor Type   HYDROMETEOR
#> 15                 Other         OTHER
#> 16               Overlay       OVERLAY
#> 17         Precipitation          PRCP
#> 18          Reflectivity  REFLECTIVITY
#> 19    Sky cover & clouds           SKY
#> 20   Spring Agricultural         SPAGR
#> 21    Spring Degree Days          SPDD
#> 22  Spring Precipitation        SPPRCP
#> 23    Spring Temperature        SPTEMP
#> 24   Summer Agricultural         SUAGR
#> 25    Summer Degree Days          SUDD
#> 
#> attr(,"class")
#> [1] "ncdc_datacats"
```

## Tornado data

The function `tornadoes()` simply gets __all the data__. So the call takes a while, but once done, is fun to play with.


```r
shp <- tornadoes()
#> OGR data source with driver: ESRI Shapefile 
#> Source: "/Users/sacmac/.rnoaa/tornadoes/tornadoes", layer: "tornado"
#> with 57988 features
#> It has 21 fields
library('sp')
plot(shp)
```

![plot of chunk unnamed-chunk-17](inst/img/unnamed-chunk-17-1.png) 

## HOMR metadata

In this example, search for metadata for a single station ID


```r
homr(qid = 'COOP:046742')
#> $`20002078`
#> $`20002078`$id
#> [1] "20002078"
#> 
#> $`20002078`$head
#>                  preferredName latitude_dec longitude_dec precision
#> 1 PASO ROBLES MUNICIPAL AP, CA      35.6697     -120.6283    DDMMSS
#>             por.beginDate por.endDate
#> 1 1949-10-05T00:00:00.000     Present
#> 
#> $`20002078`$namez
#>                         name  nameType
#> 1   PASO ROBLES MUNICIPAL AP      COOP
#> 2   PASO ROBLES MUNICIPAL AP PRINCIPAL
#> 3 PASO ROBLES MUNICIPAL ARPT       PUB
#> 
#> $`20002078`$identifiers
#>      idType          id
#> 1     GHCND USW00093209
#> 2   GHCNMLT USW00093209
#> 3      COOP      046742
#> 4      WBAN       93209
#> 5       FAA         PRB
#> 6      ICAO        KPRB
#> 7     NWSLI         PRB
#> 8 NCDCSTNID    20002078
#> 
#> $`20002078`$status
#> NULL
#> 
#> $`20002078`$platform
#> [1] "COOP"
#> 
#> $`20002078`$relocations
#>   relocation                    date
#> 1 5.1 mi NNE 1949-10-05T00:00:00.000
#> 
#> $`20002078`$remarks
#>                         type
#> 1            NWS COOP INGEST
#> 2                    GENERAL
#> 3 RIVER BASIN (COOP NETWORK)
#>                                                                                                                                                                                                                                                                                remark

NEWS
rnoaa 0.4.0
===============

NEW FEATURES

+ Gains a suite of new functions for working with NOAA GHCND data, including 
`ghcnd()`, `ghcnd_clear_cache()`, `ghcnd_countries()`, `ghcnd_search()`, `ghcnd_splitvars()`
`ghcnd_states()`, `ghcnd_stations()`, and `ghcnd_version()` (#85) (#86) (#87) (#88) (#94)
+ New contributor Adam Erickson (@DougFirErickson)
+ All NOAA buoy functions put back into the package. They were previously 
on a separate branch in the GitHub repository. (#37) (#71) (#100)

MINOR IMPROVEMENTS

+ Minor adjustments to `isd()` functions, including better man file.
+ Cleaner package imports - importing mostly only functions used in dependencies.
+ Startup message gone.
+ `callopts` parameter changed to `...` in function `swdi()`.
+ More robust test suite.
+ `ncdc()` requires that users do their own paging - previously this was done internally (#77)
+ Many dependencies dropped, simplifying package: `RCurl`, `maptools`, `stringr`, `digest`. 
A few new ones added: `dplyr`, `tidyr`.

DEPRECATED AND DEFUNCT

+ All `erddap` functions now defunct - see the package [rerddap](https://github.com/ropensci/rerddap),
a general purpose R client for ERDDAP servers. (#51) (#73) (#90) (#95)
+ The `extent` function in `noaa_stations()` used to accept either a bounding
box or a point defined by lat/long. The lat/long option dropped as it required
two packages, one of which is a pain to install for many users (#98) (#99)

rnoaa 0.3.3
===============

NEW FEATURES

+ New data source NOAA legacy API with ISD, daily, and ish data via function
`ncdc_legacy()`. (#54)
+ New function `isd()` to get ISD data from NOAA FTP server. (#76)
+ ERDDAP gridded data sets added. Now tabledap datasets are accessible via
`erddap_table()`, while gridded datasets are available via `erddap_grid()`. Helper
function `erddap_search()` was modified to search for either tabledap or griddap
datasets, and `erddap_info()` gets and prints summary information differently
for tabledap and griddap datasets. (#63)

MINOR IMPROVEMENTS

+ `erddap_data()` defunct, now as functions `erddap_table()` and `erddap_grid()`, uses new
`store` parameter which takes a function, either `disk(path, overwrite)` to store
on disk or `memory()` to store in R memory.
+ `assertthat` library removed, replaced with `stopifnot()`

rnoaa 0.3.0
===============

NEW FEATURES

+ New data source added (NOAA torndoes data) via function `tornadoes()`. (#56)
+ New data source added (NOAA storm data from IBTrACS) via functions
`storm_*()`. (#57)
+ New data source added (NOAA weather station metadata from HOMR) via functions
`homr_*()` (#59)
+ New vignettes for storm data and homr data.
+ Some functions in rnoaa now print data.frame outputs as `dplyr`-like outputs
with a summary of the data.frame, as appropriate.

MINOR IMPROVEMENTS

+ Across all `ncdc_*` functions changed `callopts` parameter to `...`. This parameter
allow you to pass in options to `httr::GET` to modify curl requests. (#61)
+ A new helper function `check_key()` looks for one of two stored keys, as an
environment variable under the name `NOAA_KEY`, or an option variable under the name
`noaakey`. Environment variables can be set during session like `Sys.setenv(VAR = "...")`,
or stored long term in your `.Renviron` file. Option variables can be set during session
like `options(var = "...")`, or stored long term in your `.Rprofile` file.
+ `is.*` and `print.*` functions no longer have public man files, but can be seen via
`rnoaa:::` if needed.

rnoaa 0.2.0
===============

NEW FEATURES

* New package imports: `sp`, `rgeos`, `assertthat`, `jsonlite`, and `ncdf4`, and new package Suggests: `knitr`, `taxize`
* Most function names changed. All `noaa*()` functions for NCDC data changed to `ncdc*()`. `noaa_buoy()` changed to `buoy()`. `noaa_seaice()` changed to `seaice()`. When you call the old versions an error is thrown, with a message pointing you to the new function name. See ?rnoaa-defunct.
* New vignettes: NCDC attributes, NCDC workflow, Seaice vignette, SWDI vignette, ERDDAP vignette, NOAA buoy vignette.
* New functions to interact with NOAA ERDDAP data: `erddap_info()`, `erddap_data()`, and `erddap_search()`.
* New functions to interact with NOAA buoy data: `buoy()`, including a number of helper functions.
* `ncdc()` now splits apart attributes. Previously, the attributes were returned as a single column, but now there is column for each attribute so data can be easily retrieved. Attribute columns differ for each different `datasetid`.
* `buoy()` function has been removed from the CRAN version of `rnoaa`. Install the version with `buoy()` and associated functions via `devtools::install_github("ropensci/rnoaa", ref="buoy")`

MINOR IMPROVEMENTS

* `noaa_swdi()` (function changed to `swdi()`) gains new parameter `filepath` to specify path to write a file to if `format=kmz` or `format=shp`. Examples added for using `format=` csv, shp, and kmz.
* Now using internal version of `plyr::compact`.
* Added API response checker/handler to all functions to pass on helpful messages on server errors.
* `ncdc()` gains new parameter `includemetadata`. If TRUE, includes metadata, if not, does not, and response should be faster as does not take time to calculate metadata.
* `noaa_stations()` gains new parameter `radius`. If `extent` is a vector of length 4 (for a bounding box) then radius is ignored, but if you pass in two points to `extent`, it is interpreted as a point, and then `radius` is used as the distance upon which to construct a bounding box. `radius` default is 10 km.

BUG FIXES

* `datasetid`, `startdate`, and `enddate` are often required parameters, and changes were made to help users with this.


rnoaa 0.1.0
===============

NEW FEATURES

* Submitted to CRAN.


rnoaa 0.0.8
===============

NEW FEATURES

* Wrote new functions for NOAA API v2.
* A working vignette now.


rnoaa 0.0.1
===============

NEW FEATURES

* Wrappers for NOAA API v1 were written, not on CRAN at this point.


R/isd.R
#' Get NOAA ISD/ISH data from NOAA FTP server.
#'
#' @export
#' @name isd
#'
#' @param usaf,wban (character) USAF and WBAN code. Required
#' @param year (numeric) One of the years from 1901 to the current year. 
#' Required.
#' @param path (character) A path to store the files, a directory. Default: 
#' \code{~/.rnoaa/isd}. Required.
#' @param overwrite (logical) To overwrite the path to store files in or not, 
#' Default: \code{TRUE}
#' @param ... Curl options passed on to \code{\link[httr]{GET}}
#' @references ftp://ftp.ncdc.noaa.gov/pub/data/noaa/
#' @details This function first looks for whether the data for your specific query has
#' already been downloaded previously in the directory given by the \code{path} 
#' parameter. If not found, the data is requested form NOAA's FTP server. The first time 
#' a dataset is pulled down we must a) download the data, b) process the data, and c) save
#' a .csv file to disk. The next time the same data is requested, we only have to read 
#' back in the .csv file, and is quite fast. The processing can take quite a long time 
#' since the data is quite messy and takes a bunch of regex to split apart text strings.
#' See examples below for different behavior.
#' @examples \dontrun{
#' # Get station table
#' stations <- isd_stations()
#' head(stations)
#'
#' # Get data
#' (res <- isd(usaf="011490", wban="99999", year=1986))
#' (res <- isd(usaf="011690", wban="99999", year=1993))
#' (res <- isd(usaf="172007", wban="99999", year=2015))
#' (res <- isd(usaf="702700", wban="00489", year=2015))
#' 
#' # The first time a dataset is requested takes longer
#' system.time( isd(usaf="782680", wban="99999", year=2011) )
#' system.time( isd(usaf="782680", wban="99999", year=2011) )
#' }

#' @export
#' @rdname isd
isd <- function(usaf, wban, year, path = "~/.rnoaa/isd", overwrite = TRUE) {
  csvpath <- isd_local(usaf, wban, year, path)
  if (!is_isd(x = csvpath)) {
    isd_GET(path, usaf, wban, year, overwrite)
  }
  message(sprintf("<path>%s", csvpath), "\n")
  structure(list(data = read_isd(csvpath, sections)), class = "isd")
}

#' @export
#' @rdname isd
isd_stations <- function(...) {
  res <- suppressWarnings(GET("ftp://ftp.ncdc.noaa.gov/pub/data/noaa/isd-history.csv", ...))
  df <- read.csv(text = content(res, "text"), header = TRUE)
  setNames(df, gsub("_$", "", gsub("\\.", "_", tolower(names(df)))))
}

#' @export
print.isd <- function(x, ..., n = 10) {
  cat("<ISD Data>", sep = "\n")
  cat(sprintf("Size: %s X %s\n", NROW(x$data), NCOL(x$data)), sep = "\n")
  trunc_mat_(x$data, n = n)
}

isd_GET <- function(bp, usaf, wban, year, overwrite) {
  dir.create(bp, showWarnings = FALSE, recursive = TRUE)
  fp <- isd_local(usaf, wban, year, bp)
  suppressWarnings(GET(isd_remote(usaf, wban, year), write_disk(fp, overwrite)))
}

isd_remote <- function(usaf, wban, year) {
  file.path(isdbase(), year, sprintf("%s-%s-%s%s", usaf, wban, year, ".gz"))
}

isd_local <- function(usaf, wban, year, path) {
  file.path(path, sprintf("%s-%s-%s%s", usaf, wban, year, ".gz"))
}

is_isd <- function(x) {
  if (file.exists(x)) TRUE else FALSE
}

isdbase <- function() 'ftp://ftp.ncdc.noaa.gov/pub/data/noaa'

read_isd <- function(x, sections) {
  path_csv <- sub("gz", "csv", x)
  if (file.exists(path_csv)) {
    df <- read.csv(path_csv, stringsAsFactors = FALSE)
  } else {
    lns <- readLines(x)
    linesproc <- lapply(lns, each_line, sections = sections)
    df <- dplyr::rbind_all(linesproc)
    cache_csv(path_csv, df)
  }
  return(df)
}

cache_csv <- function(x, y) {
  if (!file.exists(x)) {
    write.csv(y, file = x, row.names = FALSE)
  }
}

each_line <- function(y, sections){
  pluck <- function(input, x) vapply(input, "[[", numeric(1), x)
  subs <- function(z, start, stop) substring(z, start, stop)
  normal <- Map(function(a,b) subs(y, a, b), pluck(sections, "start"), pluck(sections, "stop"))
  other <- gsub("\\s+$", "", substring(y, 106, nchar(y)))
  data.frame(normal, proc_other(other), stringsAsFactors = FALSE)
}

sections <- list(
  total_chars=list(start=1,stop=4),
  usaf_station=list(start=5,stop=10),
  wban_station=list(start=11,stop=15),
  date=list(start=16,stop=23),
  time=list(start=24,stop=27),
  date_flag=list(start=28,stop=28),
  latitude=list(start=29,stop=34),
  longitude=list(start=35,stop=41),
  type_code=list(start=42,stop=46),
  elevation=list(start=47,stop=51),
  call_letter=list(start=52,stop=56),
  quality=list(start=57,stop=60),
  wind_direction=list(start=61,stop=63),
  wind_direction_quality=list(start=64,stop=64),
  wind_code=list(start=65,stop=65),
  wind_speed=list(start=66,stop=69),
  wind_speed_quality=list(start=70,stop=70),
  ceiling_height=list(start=71,stop=75),
  ceiling_height_quality=list(start=76,stop=76),
  ceiling_height_determination=list(start=77,stop=77),
  ceiling_height_cavok=list(start=78,stop=78),
  visibility_distance=list(start=79,stop=84),
  visibility_distance_quality=list(start=85,stop=85),
  visibility_code=list(start=86,stop=86),
  visibility_code_quality=list(start=87,stop=87),
  temperature=list(start=88,stop=92),
  temperature_quality=list(start=93,stop=93),
  temperature_dewpoint=list(start=94,stop=98),
  temperature_dewpoint_quality=list(start=99,stop=99),
  air_pressure=list(start=100,stop=104),
  air_pressure_quality=list(start=105,stop=105)
)

proc_other <- function(x){
  # x <- substring(x, 4, nchar(x))
  tt <- list(check_get(x, "SA1", sa1),
       check_get(x, "REM", rem),
       check_get(x, "AY1", ay1),
       check_get(x, "AY2", ay2),
       check_get(x, "AG1", ag1),
       check_get(x, "GF1", gf1),
       check_get(x, "KA1", ka1),
       check_get(x, "EQD", eqd),
       check_get(x, "MD1", md1),
       check_get(x, "MW1", mw1)
  )
  other <- tt[!sapply(tt, function(x) is.null(x[[1]]))]
  do.call(cbind, lapply(other, data.frame, stringsAsFactors = FALSE))
}

check_get <- function(string, pattern, fxn) {
  yy <- regexpr(pattern, string)
  tt <- if (yy > 0) fxn(string) else NULL
  setNames(list(tt), pattern)
}

# str_match_len(x, "SA1", 8)
str_match_len <- function(x, index, length){
  sa1 <- regexpr(index, x)
  if (sa1 > 0) {
    substring(x, sa1[1], sa1[1] + (length - 1)) 
  } else {
    NULL
  }
}

str_from_to <- function(x, a, b){
  substring(x, a, a + b)
}

str_pieces <- function(z, pieces, nms=NULL){
  tmp <- lapply(pieces, function(x) substring(z, x[1], if (x[2] == 999) nchar(z) else x[2]))
  if (is.null(nms)) tmp else setNames(tmp, nms)
}

# sea surface temperature data
# sa1(x)
sa1 <- function(x) {
  str_pieces(
    str_match_len(x, "SA1", 8),
    list(c(1,3),c(4,7),c(8,8)),
    c('sea_surface','temp','quality')
  )
}

# remarks section
# rem(x)
rem <- function(x){
  str_pieces(
    str_match_len(x, "REM", nchar(x)),
    list(c(1,3),c(4,6),c(7,9),c(10,999)),
    c('remarks','identifier','length_quantity','comment')
  )
}

# past weather manual observation
# ay1(x)
ay1 <- function(x){
  str_pieces(
    str_match_len(x, "AY1", 8),
    list(c(1,3),c(4,4),c(5,5),c(6,7),c(8,8)),
    c('manual_occurrence','condition_code','condition_quality','period','period_quality')
  )
}

# past weather manual observation
# ay2(x)
ay2 <- function(x){
  str_pieces(
    str_match_len(x, "AY2", 8),
    list(c(1,3),c(4,4),c(5,5),c(6,7),c(8,8)),
    c('manual_occurrence','condition_code','condition_quality','period','period_quality')
  )
}

# PRECIPITATION-ESTIMATED-OBSERVATION identifier
# ag1(x)
ag1 <- function(x){
  str_pieces(
    str_match_len(x, "AG1", 7),
    list(c(1,3),c(4,4),c(5,7)),
    c('precipitation','discrepancy','est_water_depth')
  )
}

# sky condition
# gf1(x)
gf1 <- function(x){
  str_pieces(
    str_match_len(x, "GF1", 26),
    list(c(1,3),c(4,5),c(6,7),c(8,8),c(9,10),c(11,11),c(12,13),c(14,14),c(15,19),c(20,20),c(21,22),c(23,23),c(24,25),c(26,26)),
    c('sky_condition','coverage','opaque_coverage','coverage_quality','lowest_cover','lowest_cover_quality',
      'low_cloud_genus','low_cloud_genus_quality','lowest_cloud_base_height','lowest_cloud_base_height_quality',
      'mid_cloud_genus','mid_cloud_genus_quality','high_cloud_genus','high_cloud_genus_quality')
  )
}

# extreme air temperature
# ka1(x)
ka1 <- function(x){
  str_pieces(
    str_match_len(x, "KA1", 13),
    list(c(1,3),c(4,6),c(7,7),c(8,12),c(13,13)),
    c('extreme_temp','period_quantity','max_min','temp','temp_quality')
  )
}

# element data quality section
# eqd(x)
eqd <- function(x){
  eqdtmp <- str_match_len(x, "EQD", nchar(x))
  eqdmtchs <- gregexpr("Q[0-9]{2}", eqdtmp)
  segments <- str_from_to(eqdtmp, eqdmtchs[[1]], 13)
  lapply(segments, function(m){
    str_pieces(m,
               list(c(1,3),c(4,9),c(10,10),c(11,16)),
               c('observation_identifier','observation_text','reason_code','parameter')
    )
  })
}

# atmospheric pressure change
# md1(x)
md1 <- function(x){
  str_pieces(
    str_match_len(x, "MD1", 14),
    list(c(1,3),c(4,4),c(5,5),c(6,8),c(9,9),c(10,13),c(14,14)),
    c('atmospheric_change','tendency','tendency_quality','three_hr','three_hr_quality',
      'twentyfour_hr','twentyfour_hr_quality')
  )
}

# PRESENT-WEATHER-OBSERVATION manual occurrence identifier, MW1=first weather reported
# mw1(x)
mw1 <- function(x){
  str_pieces(
    str_match_len(x, "MW1", 6),
    list(c(1,3),c(4,5),c(6,6)),
    c('first_weather_reported','condition','condition_quality')
  )
}
